src/polynomial.cpp

```diff
@@ -0,0 +1,307 @@
+#include "polynomial.h"
+#include "add.h"
+#include "mul.h"
+#include "pow.h"
+#include "constants.h"
+
+namespace SymEngine {
+
+Polynomial::Polynomial(const RCP<const Symbol> &var, map_uint_mpz&& dict) :
+     var_{var}, dict_{std::move(dict)} {
+
+    map_uint_mpz::iterator it = dict_.begin();
+    uint largest = it->first;
+    uint cur;
+    while (it != dict_.end()) {
+        cur =it->first;
+        if (cur > largest)
+            largest = cur;
+        it++;
+    }
+    degree = largest;
```

Comment:
This is calculating the degree of the polynomial, right? I think this should be passed in as a parameter. Many time the user (or the caller) knows the degree, so you just want to pass it in. Check in `is_canonical` that the degree is equal to the calculated one. That way the constructor is very fast in Release mode.